pkg/cluster/cluster.go

```diff
@@ -208,37 +190,56 @@ func (c *DynamicCluster) getNodeForKey(partitionKey string) (string, bool) {
 	c.refreshRing()
 	c.lock.Lock()
 	defer c.lock.Unlock()
+
 	return c.ring.GetNode(partitionKey)
 }
 
+// Join attempts to join a cluster by contacting all the given seed hosts.
+//
+// This simply delegates to github.com/hashicorp/memberlist#Memberlist.Join.
 func (c *DynamicCluster) Join(seeds []string) error {
 	_, err := c.ml.Join(seeds)
 	return err
 }
 
+// NotifyJoin is the callback that is invoked when a node is detected to have joined.
+//
+// See github.com/hashicorp/memberlist#EventDelegate.NotifyJoin
 func (c *DynamicCluster) NotifyJoin(node *memberlist.Node) {
 	c.lock.Lock()
 	defer c.lock.Unlock()
-	if c.ml == nil {
-		c.ring = hashring.New([]string{node.Name})
-		return
-	}
 	log.Printf("[%s] node joined %s", c.name, node.Name)
 	c.ring.AddNode(node.Name)
 }
 
+// NotifyLeave is the callback that is invoked when a node is detected to have left.
+//
+// See github.com/hashicorp/memberlist#EventDelegate.NotifyLeave
 func (c *DynamicCluster) NotifyLeave(node *memberlist.Node) {
 	c.lock.Lock()
 	defer c.lock.Unlock()
 	log.Printf("[%s] node left %s", c.name, node.Name)
 	c.ring.RemoveNode(node.Name)
 }
+
+// NotifyUpdate is the callback that is invoked when a node to have updated.
```

Comment:
there seems to be a type in this new doc here. Should be `node is detected to have been updated`